transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java

```diff
@@ -206,7 +207,42 @@ protected void doWrite(ChannelOutboundBuffer in) throws Exception {
                     // Break the loop and so incompleteWrite(...) is called.
                     break;
                 }
+            } else if (msg instanceof AbstractFileRegion) {
+                AbstractFileRegion region = (AbstractFileRegion) msg;
+                long transferableBytes = region.transferableBytes();
+                if (transferableBytes == 0) {
+                    in.remove();
+                    continue;
+                }
+                boolean done = false;
+                long flushedAmount = 0;
+                if (writeSpinCount == -1) {
+                    writeSpinCount = config().getWriteSpinCount();
+                }
+
+                for (int i = writeSpinCount - 1; i >= 0; i--) {
+                    long localFlushedAmount = doWriteFileRegion(region);
+                    if (localFlushedAmount == 0) {
+                        setOpWrite = true;
+                        break;
+                    }
+
+                    flushedAmount += localFlushedAmount;
+                    if (!region.isTransferable()) {
+                        done = true;
+                        break;
+                    }
+                }
+                in.progress(flushedAmount);
+                if (done) {
+                    in.remove();
+                } else {
+                    // Break the loop and so incompleteWrite(...) is called.
+                    break;
+                }
             } else if (msg instanceof FileRegion) {
+                // TODO: consider remove this branch once we move the methods in AbstractFileRegion
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)